clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java

```diff
@@ -151,6 +158,93 @@ public void testSslPasswords() {
         assertEquals(Password.HIDDEN, vals.get(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG).toString());
     }
 
+    @Test
+    public void testValidate() {
+        Map<String, Config> expected = new HashMap<>();
+        String errorMessageB = "Missing required configuration \"b\" which has no default value.";
+        String errorMessageC = "Missing required configuration \"c\" which has no default value.";
+        String errorMessageD = "Invalid value for configuration d";
+
+        Config configA = new Config("a", 1, Arrays.<Object>asList(1, 2, 3), Collections.<String>emptyList());
+        Config
+            configB = new Config("b", null, Arrays.<Object>asList(4, 5), Arrays.asList(errorMessageB, errorMessageB));
+        Config
+            configC = new Config("c", null, Arrays.<Object>asList(4, 5), Arrays.asList(errorMessageC));
+        Config
+            configD = new Config("d", 10, Arrays.<Object>asList(1, 2, 3), Arrays.asList(errorMessageD));
+        expected.put("a", configA);
+        expected.put("b", configB);
+        expected.put("c", configC);
+        expected.put("d", configD);
+
+        ConfigDef def = new ConfigDef()
+            .define("a", Type.INT, Importance.HIGH, "docs", "group", 1, Width.SHORT, "a", Arrays.asList("b", "c"), new IntegerRecommnder())
+            .define("b", Type.INT, Importance.HIGH, "docs", "group", 2, Width.SHORT, "b", Collections.<String>emptyList(), new IntegerRecommnder())
+            .define("c", Type.INT, Importance.HIGH, "docs", "group", 3, Width.SHORT, "c", Collections.<String>emptyList(), new IntegerRecommnder())
+            .define("d", Type.INT, Importance.HIGH, "docs", "group", 4, Width.SHORT, "d", Collections.singletonList("b"), new IntegerRecommnder());
+
+        Map<String, String> props = new HashMap<>();
+        props.put("a", "1");
+        props.put("d", "10");
+
+        List<Config> configs = def.validate(props);
+        for (Config config : configs) {
+            String name = config.getName();
+            Config expectedConfig = expected.get(name);
+            assertTrue(compareConfigValues(expectedConfig, config));
+        }
+    }
+
+    private boolean compareConfigValues(Config a, Config b) {
```

Comment:
You might want to rename this `assertConfigValuesEqual` and do assertions in here. Otherwise the actual comparison that failed is a pain to track down. Also makes it clearer what's going on in the test code that calls this method.